tests/integration/pilot/common/routing.go

```diff
@@ -855,77 +855,41 @@ func selfCallsCases(apps *EchoDeployments) []TrafficTestCase {
-			}
-
-			for _, destinations := range destinationSets {
-				if len(destinations) == 0 {
-					continue
-				}
-				client := client
-				destinations := destinations
-				// grabbing the 0th assumes all echos in destinations have the same service name
-				destination := destinations[0]
-				if apps.Headless.Contains(destination) && len(apps.Headless) > 1 {
-					// TODO(landow) incompatibilities with multicluster & headless
-					continue
-				}
-				if apps.Naked.Contains(client) && apps.VM.Contains(destination) {
-					// Need a sidecar to connect to VMs
-					continue
-				}
-				if apps.VM.Contains(client) && apps.External.Contains(destination) {
-					// TODO VM won't resolve DNS in ServiceEntry https://github.com/istio/istio/issues/27154
-					continue
-				}
 
-				type protocolCase struct {
-					// The port we call
-					port string
-					// The actual type of traffic we send to the port
-					scheme scheme.Instance
-				}
-				protocols := []protocolCase{
-					{"http", scheme.HTTP},
-					{"auto-http", scheme.HTTP},
-					{"tcp", scheme.TCP},
-					{"auto-tcp", scheme.TCP},
-					{"grpc", scheme.GRPC},
-					{"auto-grpc", scheme.GRPC},
-				}
+	type protocolCase struct {
+		// The port we call
+		port string
+		// The actual type of traffic we send to the port
+		scheme scheme.Instance
+	}
+	protocols := []protocolCase{
+		{"http", scheme.HTTP},
+		{"auto-http", scheme.HTTP},
+		{"tcp", scheme.TCP},
+		{"auto-tcp", scheme.TCP},
+		{"grpc", scheme.GRPC},
+		{"auto-grpc", scheme.GRPC},
+	}
 
-				// so we can validate all clusters are hit
-				callCount := callsPerCluster * len(destinations)
-				for _, call := range protocols {
-					call := call
-					cases = append(cases, TrafficTestCase{
-						// TODO(https://github.com/istio/istio/issues/26798) enable sniffing tcp
-						skip: call.scheme == scheme.TCP,
-						name: fmt.Sprintf("%v %v->%v from %s", call.port, client.Config().Service, destination.Config().Service, client.Config().Cluster.StableName()),
-						call: client.CallWithRetryOrFail,
-						opts: echo.CallOptions{
-							Target:   destination,
-							PortName: call.port,
-							Scheme:   call.scheme,
-							Count:    callCount,
-							Timeout:  time.Second * 5,
-							Validator: echo.And(
-								echo.ExpectOK(),
-								echo.ExpectHost(destination.Config().HostHeader())),
-						},
-					})
-				}
-			}
-		}
+	// so we can validate all clusters are hit
+	for _, call := range protocols {
+		call := call
+		cases = append(cases, TrafficTestCase{
+			// TODO(https://github.com/istio/istio/issues/26798) enable sniffing tcp
+			skip: call.scheme == scheme.TCP,
+			name: call.port,
+			opts: echo.CallOptions{
+				PortName: call.port,
+				Scheme:   call.scheme,
+				Timeout:  time.Second * 5,
+			},
+			validate: func(src echo.Instance, dst echo.Instances) echo.Validator {
+				return echo.And(
+					echo.ExpectOK(),
+					echo.ExpectHost(dst[0].Config().HostHeader()))
+			},
+			workloadAgnostic: true,
```

Comment:
do we want to just stop using TrafficTestCase and use echotest directly?

Reply:
I guess there is still some boilerplate

Reply:
TrafficTestCase might deserve a package or something that helps consolidate with the security reachability tests.